language/move-lang/src/cfgir/inline_blocks.rs

```diff
@@ -0,0 +1,90 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use super::cfg::BlockCFG;
+use crate::hlir::ast::{BasicBlocks, Command_, Label};
+use std::collections::{BTreeMap, BTreeSet};
+
+/// returns true if anything changed
+pub fn optimize(cfg: &mut BlockCFG) -> bool {
+    let changed = optimize_(cfg.blocks_mut());
+    if changed {
+        let dead_blocks = cfg.recompute();
+        assert!(dead_blocks.is_empty())
+    }
+    changed
+}
+
+fn optimize_(blocks: &mut BasicBlocks) -> bool {
+    let single_target_labels = find_single_target_labels(&blocks);
+    inline_single_target_blocks(&single_target_labels, blocks)
+}
+
+fn find_single_target_labels(blocks: &BasicBlocks) -> BTreeSet<Label> {
+    use Command_ as C;
+    let mut counts = BTreeMap::new();
+    for block in blocks.values() {
+        match &block.back().unwrap().value {
+            C::JumpIf {
+                cond: _cond,
+                if_true,
+                if_false,
+            } => {
+                *counts.entry(*if_true).or_insert(0) += 1;
+                *counts.entry(*if_false).or_insert(0) += 1
+            }
+            C::Jump(lbl) => *counts.entry(*lbl).or_insert(0) += 1,
+            _ => (),
+        }
+    }
+    counts
+        .into_iter()
+        .filter(|(_, count)| *count == 1)
+        .map(|(lbl, _)| lbl)
+        .collect()
+}
+
+fn inline_single_target_blocks(
+    single_jump_targets: &BTreeSet<Label>,
+    blocks: &mut BasicBlocks,
+) -> bool {
+    let labels_vec = blocks.keys().cloned().collect::<Vec<_>>();
+    let mut labels = labels_vec.into_iter();
+    let mut next = labels.next();
+
+    let mut working_blocks = std::mem::replace(blocks, BasicBlocks::new());
+    let finished_blocks = blocks;
+
+    let mut changed = false;
+    while let Some(cur) = next {
+        let mut block = match working_blocks.remove(&cur) {
+            None => {
+                next = labels.next();
+                continue;
+            }
+            Some(b) => b,
+        };
+
+        match block.back().unwrap() {
+            // check for single target forward jump
```

Comment:
Maybe checking for forward jumps is safe -- I'm not sure -- but it seems like the wrong condition to check. It seems to me like inlining should be fine for a backward jump as long as there are no other jumps to that label and if there is no fall-through into that block. You're already checking for the single-jump condition. How about changing to check for fall-through instead of forward jumps?

Reply:
There is are no fall thoughs in the cfgir. And the block ordering might not be contiguous, so you cannot check for `cur.0 + 1 == target.0` 

The point of this check is to ensure that there are no loops. The backward jumps are *not* safe if there are loops. And currently the *only* back jumps we emit are loops

That being said. I believe that in the current system, if it is not the start block, loop heads always serve as the target for 2 or more jumps. So I will switch the check for not being the start

Reply:
Ah, OK, I didn't realize that the cfgir has no fallthroughs. Maybe I'm being dense, but what's the problem with backward jumps with loops?

Reply:
Imagine this scenario 

```
label Start:
    JumpIf(copy(check),  End, Next)

Reply:
In that case, there would be a fall-through into InfiniteLoopHead that would block the inlining. (Either that or a jump into the body of the loop, but you're going to ban that with the pending control flow checks.)

Reply:
Sorry, not a fall-through, but another jump to the loop head (which is what you said above). It sounds like we're in agreement on that.